common/utils/cpp_circular_buffer.h

```diff
@@ -156,6 +156,60 @@ class CircularBuffer {
    * @return bool: True if buffer is full, false otherwise.
    */
   bool full() const { return this->count() == Capacity; }
+
+  /**
+   * Get the average of all elements in the buffer.
+   * 
+   * @return Type: The average of all elements in the buffer.
+   */
+  Type average() {
+    Type sum = this->sum();
+    return sum / this->count();
+  }
+
+  /**
+   * Get the median of all elements in the buffer.
+   * 
+   * @return Type: The median of all elements in the buffer.
+   */
+  Type median() {
+    size_t size = this->count();
+    std::vector<Type> linearData(size);
+
+    for (size_t i = 0; i < size; i++) {
+      linearData[i] = getNthValue(i);
+    }
+
+    std::sort(linearData.begin(), linearData.end());
+    return linearData[size / 2];
+  }
+
+  /**
+   * Get the n'th elements in the buffer.
+   * 
+   * @return Type: The n'th element in the buffer.
+   */
+  Type getNthValue(int n) {
```

Comment:
should there be any checks to see if n is less than 0 or bigger than this->size()?

Reply:
Yeah, that would be good. But then you need to either return optional, pass the return value by reference, or return Type{0}.

Reply:
For context, Matt means an `std::optional` which essentially has added information about whether or not it holds a value, so if the `getNthValue` call can't use the given `n`, it has something to return. See https://en.cppreference.com/w/cpp/utility/optional

Reply:
Also, I think it would be good if this could return a reference or pointer rather than a full `Type`. We might hold a big struct or something in the buffer that we don't want to make a copy of when someone asks for it.

Also also, look into `operator ()` overloading which can provide a potentially easier way of doing something like this function (you can have both, though, if you want). We do this in https://github.com/AerospaceNU/stm32-avionics/blob/main/common/utils/fcb_matrix.h

Reply:
added a destination pointer like the peek method does and changed the return type to boolean.